### Financial Fraud 
# Objective

The goal of this notebook is to understand the data and finally, perform EDA and feature engineering for the future ML model

In [36]:
import numpy as np
import pandas as pd
import matplotlib as plt
import seaborn as sns

In [37]:
df = pd.read_csv('Data/Fraud dataset.csv')

In [38]:
df.head()

,step,type,amount,nameOrig,oldbalanceOrg,newbalanceOrig,nameDest,oldbalanceDest,newbalanceDest,isFraud,isFlaggedFraud
0,1,PAYMENT,9839.64,C1231006815,170136.0,160296.36,M1979787155,0.0,0.0,0,0
1,1,PAYMENT,1864.28,C1666544295,21249.0,19384.72,M2044282225,0.0,0.0,0,0
2,1,TRANSFER,181.00,C1305486145,181.0,0.00,C553264065,0.0,0.0,1,0
3,1,CASH_OUT,181.00,C840083671,181.0,0.00,C38997010,21182.0,0.0,1,0
4,1,PAYMENT,11668.14,C2048537720,41554.0,29885.86,M1230701703,0.0,0.0,0,0


In [39]:
df.tail()

,step,type,amount,nameOrig,oldbalanceOrg,newbalanceOrig,nameDest,oldbalanceDest,newbalanceDest,isFraud,isFlaggedFraud
6362615,743,CASH_OUT,339682.13,C786484425,339682.13,0.0,C776919290,0.00,339682.13,1,0
6362616,743,TRANSFER,6311409.28,C1529008245,6311409.28,0.0,C1881841831,0.00,0.00,1,0
6362617,743,CASH_OUT,6311409.28,C1162922333,6311409.28,0.0,C1365125890,68488.84,6379898.11,1,0
6362618,743,TRANSFER,850002.52,C1685995037,850002.52,0.0,C2080388513,0.00,0.00,1,0
6362619,743,CASH_OUT,850002.52,C1280323807,850002.52,0.0,C873221189,6510099.11,7360101.63,1,0


In [40]:
df.sample(5)

,step,type,amount,nameOrig,oldbalanceOrg,newbalanceOrig,nameDest,oldbalanceDest,newbalanceDest,isFraud,isFlaggedFraud
5791280,401,PAYMENT,27786.01,C649301514,0.00,0.00,M2118381354,0.00,0.00,0,0
275684,15,CASH_OUT,394553.50,C1301539611,908385.31,513831.81,C1820415853,995332.44,1449082.08,0,0
5007893,353,TRANSFER,1216510.70,C1041723354,19920.00,0.00,C2038045635,37572.44,1254083.14,0,0
5930185,404,PAYMENT,5713.13,C286879848,757.00,0.00,M1850052769,0.00,0.00,0,0
1304360,136,PAYMENT,17681.81,C28526041,200374.00,182692.19,M1509514591,0.00,0.00,0,0


In [41]:
df.shape

(6362620, 11)

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            object 
 2   amount          float64
 3   nameOrig        object 
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        object 
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 534.0+ MB


### Features description

| Features | Data Type | Description | Category |
|:-----|:------:|:------:|:------:|
| step | Integer |Time step of the transaction (each step represents one hour in the simulation) |  Time |
| type | Object |Type of transaction | Categorical |
| amount | Integer |Amount involved in the transaction | Numerical |
| nameOrig | Object |Unique identifier of the customer initiating the transaction | Identifier |
| oldbalanceOrig | Float | Sender's account balance before the transaction | Numerical |
| newbalanceOrig | Float | Sender's account balance after the transaction | Numerical |
| nameDest | Object | Unique identifier of the receiving account | Identifier |
| oldbalanceDest | Float | Receiver's account balance before the transaction | Numerical |
| newbalanceDest | Float | Receiver's account balance before the transaction | Numerical |
| isFraud | Integer | Target variable indicating whether the transaction is fraudulent (1) or legitimate (0) | Target |
| isFlaggedFraud | Integer | Indicates whether the transaction was flagged by the bank's rule-based fraud detection system | Target |

### Feature Categories

**Time Feature**
- step
- 
**Numerical Features**
- amount
- oldbalanceOrg
- newbalanceOrig
- oldbalanceDest
- newbalanceDest

**Categorical Feature**
- type

**Identifier Features**
- nameOrig
- nameDest

**Target Variable**
- isFraud

**Bank Flag (Not the prediction target)**
- isFlaggedFraud

In [54]:
df.describe(include="all")

,step,type,amount,nameOrig,oldbalanceOrg,newbalanceOrig,nameDest,oldbalanceDest,newbalanceDest,isFraud,isFlaggedFraud
count,6.362620e+06,6362620,6.362620e+06,6362620,6.362620e+06,6.362620e+06,6362620,6.362620e+06,6.362620e+06,6.362620e+06,6.362620e+06
unique,NaN,5,NaN,6353307,NaN,NaN,2722362,NaN,NaN,NaN,NaN
top,NaN,CASH_OUT,NaN,C1902386530,NaN,NaN,C1286084959,NaN,NaN,NaN,NaN
freq,NaN,2237500,NaN,3,NaN,NaN,113,NaN,NaN,NaN,NaN
mean,2.433972e+02,NaN,1.798619e+05,NaN,8.338831e+05,8.551137e+05,NaN,1.100702e+06,1.224996e+06,1.290820e-03,2.514687e-06
std,1.423320e+02,NaN,6.038582e+05,NaN,2.888243e+06,2.924049e+06,NaN,3.399180e+06,3.674129e+06,3.590480e-02,1.585775e-03
min,1.000000e+00,NaN,0.000000e+00,NaN,0.000000e+00,0.000000e+00,NaN,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,1.560000e+02,NaN,1.338957e+04,NaN,0.000000e+00,0.000000e+00,NaN,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
50%,2.390000e+02,NaN,7.487194e+04,NaN,1.420800e+04,0.000000e+00,NaN,1.327057e+05,2.146614e+05,0.000000e+00,0.000000e+00
75%,3.350000e+02,NaN,2.087215e+05,NaN,1.073152e+05,1.442584e+05,NaN,9.430367e+05,1.111909e+06,0.000000e+00,0.000000e+00


In [55]:
df.isnull().sum()

step              0
type              0
amount            0
nameOrig          0
oldbalanceOrg     0
newbalanceOrig    0
nameDest          0
oldbalanceDest    0
newbalanceDest    0
isFraud           0
isFlaggedFraud    0
dtype: int64

In [56]:
df.duplicated().sum()

np.int64(0)

In [57]:
df.nunique()

step                  743
type                    5
amount            5316900
nameOrig          6353307
oldbalanceOrg     1845844
newbalanceOrig    2682586
nameDest          2722362
oldbalanceDest    3614697
newbalanceDest    3555499
isFraud                 2
isFlaggedFraud          2
dtype: int64

In [58]:
df[df['type'] == 'TRANSFER'].sample(5)

,step,type,amount,nameOrig,oldbalanceOrg,newbalanceOrig,nameDest,oldbalanceDest,newbalanceDest,isFraud,isFlaggedFraud
6111145,525,TRANSFER,79289.82,C751029393,0.0,0.0,C2128476391,1660684.96,1739974.79,0,0
4682743,331,TRANSFER,199181.82,C204884370,0.0,0.0,C775186040,294927.53,494109.35,0,0
3643369,275,TRANSFER,145973.36,C714592447,24752.0,0.0,C952107216,0.00,145973.36,0,0
4291080,307,TRANSFER,278864.50,C15240734,13158.0,0.0,C1632658389,156927.03,435791.53,0,0
3692608,277,TRANSFER,352534.85,C1476781823,10738.0,0.0,C2084447034,4194.20,356729.04,0,0


In [59]:
df['isFraud'].value_counts()

isFraud
0    6354407
1       8213
Name: count, dtype: int64

In [60]:
df["isFraud"].value_counts(normalize=True) * 100

isFraud
0    99.870918
1     0.129082
Name: proportion, dtype: float64

In [61]:
df['type'].value_counts()

type
CASH_OUT    2237500
PAYMENT     2151495
CASH_IN     1399284
TRANSFER     532909
DEBIT         41432
Name: count, dtype: int64

In [62]:
df.nunique()

step                  743
type                    5
amount            5316900
nameOrig          6353307
oldbalanceOrg     1845844
newbalanceOrig    2682586
nameDest          2722362
oldbalanceDest    3614697
newbalanceDest    3555499
isFraud                 2
isFlaggedFraud          2
dtype: int64

### Key Findings

- Sender's Balance sheet is wrong, as ((oldbalanceOrg - newbalanceOrg) != Tranfer)

**Target Variable:** 
-

**1. Target Variable:**
- isFraud

**2. Missing Values:**
Altough there is no missing value,  but there are some wrong values in -
- oldbalanceOrg 
- newbalanceOrig

**3. Duplicate Rows:**
No Missing rows.

**4. Fraud Percentage:**
0.129082%

**5. Numerical Features:**
- amount
- oldbalanceOrg
- newbalanceOrig
- oldbalanceDest
- newbalanceDest

**6. Categorical Features:**
- type

**7. Potential Challenges:**
- Highly imbalanced dataset
- possible data leakage from isFlaggedFraud

**8. Possible Feature Engineering Ideas:**
- Balance differnces
- Transaction Time (Hours and Day)
- 